### DISCUSSION TITLE
NB: When using SQLModel + Alembic, make sure ForeignKeys are named!

### DISCUSSION THREAD

**Discussion by thomasborgen** [fastapi/sqlmodel#1213] [Show and tell]
When using alembic to create migrations from sqlmodel Alembic does not automatically add names to ForeignKey constraints and other constraints. 

## TLDR

Setup sqlalchemy Metadata with naming conventions so that foreign keys and other constraints can be referred to in later migrations. Otherwise many FKs can't be changed or removed.

```python
from pydantic import ConfigDict
from sqlmodel import Field, SQLModel

convention = {
    "ix": "ix_%(column_0_label)s",
    "uq": "uq_%(table_name)s_%(column_0_name)s",
    "ck": "ck_%(table_name)s_%(constraint_name)s",
    "fk": "fk_%(table_name)s_%(column_0_name)s_%(referred_table_name)s",
    "pk": "pk_%(table_name)s",
}


class SimpleModel(SQLModel):
    """Base model for everything."""

    model_config = ConfigDict(
        protected_namespaces=(),
    )  # type: ignore


SimpleModel.metadata.naming_convention = convention
```

## Longer version

Why is this important?

Well if you have a model with more than 1 FK like 

```python
class ProductBase(SimpleModel):
    id: uuid.UUID = Field(default_factory=uuid.uuid4, primary_key=True)
    variant_id: uuid.UUID = Field(foreign_key="variant.id")
    parent_id: uuid.UUID | None = Field(foreign_key="product.id")

class Product(ProductBase, table=True):
     # Foreign key relationships
    variant: Variant = Relationship(back_populates="products")
    parent: Optional["Product"] = Relationship(
        back_populates="children",
        sa_relationship_kwargs={
            "remote_side": "Product.id"  # Refers to the table Class
        },
    )
    # Reverse populations
    children: list["Product"] = Relationship(back_populates="parent")
```

This creates an alembic migration that looks something like:

```python
    op.create_table(
        "product",
        sa.Column("id", sa.Uuid(), nullable=False),
        sa.Column("variant_id", sa.Uuid(), nullable=False),
        sa.Column("parent_id", sa.Uuid(), nullable=True),
        sa.ForeignKeyConstraint(
            ["parent_id"],
            ["product.id"],
        ),
        sa.ForeignKeyConstraint(
            ["variant_id"],
            ["variant.id"],
        ),
        sa.PrimaryKeyConstraint("id"),
    )
```

`ForeignKeyConstraint` has a keyword argument `name: _ConstraintNameArgument = None,` This is not set by default in SQLModel models.

So if I want to change a FK or even if i want to remove a FK **i can't!!** Because there is no name sql can refer to when deleting it.

This can cause so many issues down the line. Fix is in the TLDR, **add sqlalchemy naming conventions**.

Given how userfriendly SQLModel is and how easy it is to work with because it hides a lot of complexity I think that SQLModel should define these names by default but also make it easy to set up custom ones with `ConfigDict()`. Maybe having an argument in `Relationship` called `name` could also make sense.

**Comment by chipfranzen** [comment]
Just ran into this issue. Seems like something that should be baked into SQLModel, no?

**Reply by thomasborgen** [reply]
To be fair, it should be baked in to alembic. Alembic needs named foreign keys and named constraints to alter foreign keys and constraints. So Alembic should make sure they exist. . 🤷🏼
